### PR TITLE
terminal/command: Add 'reload' command

### LINKED ISSUE #1551: Feature: Add reload command to refresh code being debugged
1. What version of Delve are you using (`dlv version`)? 1.2.0
2. What version of Go are you using? (`go version`)? 1.12.1 darwin/amd64
3. What operating system and processor architecture are you using? OS x/intel
4. What did you do? I ran delve to debug an issue and set some breakpoints, found the issue and updated my source code.  Afterwards, in the same delve instance, I issued a restart command
5. What did you expect to see? N/A
6. What did you see instead? new source files and old code running.

The current behavior is similar to how gdb works.  After you edit the code, gdb will show the new source files but run old code.  This can be handy if you want to avoid recompiling because your disk is excessively slow.

However, in gdb you can issue a ```file "executable.exe"``` command.  At this point it reloads the executable from disk and you now have updated machine code.  You can issue a ```restart``` and verify the changes you made are working as expected, with your existing breakpoints and traces still in place.

The ```file``` command wouldn't make sense for delve because of delves ability to run the code without an executable file on disk.  How about adding a ```reload``` command that triggers delve to read sources from disk and recompile?  Is this sort of thing possible? hard? easy? foolhardy?

### PR DESCRIPTION
**Pull request by alexsaezm**
These changes add the 'reload' command, which allows us to rebuild the project
and start the debugging session again. Currently, if the project's code is
updated while debugging it, Delve shows the new source code, but it's still
running the old one. With 'reload', the whole binary is rebuilt, and the
process starts again.

Fixes #1551

### CODE REVIEW

File: service/debugger/debugger.go [hunk 1]
```diff
@@ -420,6 +428,12 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 		p, err = d.recordingRun(run)
 		d.recordingDone()
 	} else {
+		if rebuild {
```
Comment (aarzilli) [inline service/debugger/debugger.go]: This should go before the `if recorded` line, otherwise it won't have an effect on recorded targets.

Also, breakpoints that can't be set by (file, line) should be discarded in the `for _, oldBp := ...` loop below.

File: service/debugger/debugger.go [hunk 2]
```diff
@@ -420,6 +428,12 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
...
+			err := gobuild.GoBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
```
Comment (aarzilli) [inline service/debugger/debugger.go]: Some error message should be returned if the debugger was started with `attach` or `exec`, also this doesn't work if the debugger was started with `test`.

File: pkg/terminal/command.go [hunk 3]
```diff
@@ -142,6 +142,7 @@ For live targets the command takes the following forms:
 If newargv is omitted the process is restarted (or re-recorded) with the same argument vector.
 If -noargs is specified instead, the argument vector is cleared.
 `},
+		{aliases: []string{"reload"}, group: runCmds, cmdFn: c.reload, allowedPrefixes: revPrefix, helpMsg: "Refresh code."},
```
Comment (aarzilli) [inline pkg/terminal/command.go]: I think `rebuild` would be a better name.

File: cmd/dlv/cmds/commands.go [hunk 4]
```diff
@@ -47,6 +47,10 @@ var (
 	InitFile string
 	// BuildFlags is the flags passed during compiler invocation.
 	BuildFlags string
+	// DlvArgs DlvArgs contains debug arguments.
+	DlvArgs []string
```
Comment (derekparker) [inline cmd/dlv/cmds/commands.go]: I don't think we need to add these here and export them.

File: cmd/dlv/cmds/commands.go [hunk 5]
```diff
@@ -477,20 +481,20 @@ func traceCmd(cmd *cobra.Command, args []string) {
...
+					if err := gobuild.GoBuild(debugname, DlvArgs, BuildFlags); err != nil {
 						fmt.Fprintf(os.Stderr, "%v\n", err)
 						return 1
 					}
 				}
 				defer gobuild.Remove(debugname)
 			}
 
-			processArgs = append([]string{debugname}, targetArgs...)
+			processArgs = append([]string{debugname}, TargetArgs...)
```
Comment (derekparker) [inline cmd/dlv/cmds/commands.go]: (I can't add a review comment on the appropriate line cause it wasn't changed so I'll add it here)

You must fill in the new struct members you added to `service.Config` below (`Packages` and `BuildFlags`). Currently they are not being passed through and are always empty, it appears.

**Comment by alexsaezm** [comment]
I added a test for the feature, and also move the `ExecuteKind` from `cmds` to `debugger`. Also, I renamed the feature because yeah, makes total sense to call it rebuild.

File: service/debugger/debugger.go [hunk 6]
```diff
@@ -434,6 +434,15 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 	}
 	var p *proc.Target
 	var err error
+
+	// TODO executeKind from command.go
+	if rebuild && d.canRestart() {
```
Comment (aarzilli) [inline service/debugger/debugger.go:443]: The code above has a `if !d.canRestart() { return ... }` why do you need to check this here too?

Also the loop that sets breakpoints below should discard the ones that can't be set by file/line.

File: go.mod [hunk 7]
```diff
@@ -20,6 +20,6 @@ require (
 	go.starlark.net v0.0.0-20190702223751-32f345186213
 	golang.org/x/arch v0.0.0-20190927153633-4e8777c89be4
 	golang.org/x/sys v0.0.0-20190626221950-04f50cda93cb
-	golang.org/x/tools v0.0.0-20191127201027-ecd32218bd7f
+	golang.org/x/tools v0.0.0-20200415034506-5d8e1897c761
```
Comment (aarzilli) [inline go.mod:23]: I don't think this needs an updated version of golang.org/x/tools so this should be reverted. If I'm wrong and it does the vendor directory should also be updated.

Comment (alexsaezm) [reply go.mod:23]: Yeah, for sure it is not necessary. My bad, I think I update this while doing the rebase of the branch for some reason.

File: service/debugger/debugger.go (hunk 6 above)
Comment (alexsaezm) [reply service/debugger/debugger.go]: You are totally right. I removed the d.canRestart.

Can you please clarify a little more the loop part? I don't really understand what you mean.

Comment (aarzilli) [reply service/debugger/debugger.go]: In this function there's a loop, `for _, oldBp := ...` that restores breakpoints. It either does it by file:line (`if len(oldBp).File > 0`) or by address (`else`). We shouldn't try to set them by address because addresses have changed after rebuilding.

File: service/debugger/debugger.go [hunk 8]
```diff
@@ -420,6 +439,20 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 	}
 	var p *proc.Target
 	var err error
+
+	if rebuild {
+		// We cannot restart processes that we don't know how to build. If
+		// d.config.Kind is different from ExecutingGeneratedFile that means we
+		// didn't build the binary and we are just running it
+		if d.config.Kind == ExecutingExistingFile {
+			return nil, fmt.Errorf("cannot restart process Delve did not create")
```
Comment (aarzilli) [inline service/debugger/debugger.go:448]: s/cannot restart/cannot rebuild/

File: cmd/dlv/cmds/commands.go [hunk 9]
```diff
@@ -387,11 +387,11 @@ func dapCmd(cmd *cobra.Command, args []string) {
 		if workingDir != "" {
 			fmt.Fprintf(os.Stderr, "Warning: working directory ignored with dap; launch requests must specify full program path\n")
 		}
-		dlvArgs, targetArgs := splitArgs(cmd, args)
-		if len(dlvArgs) > 0 {
+		DlvArgs, TargetArgs := splitArgs(cmd, args)
```
Comment (aarzilli) [inline cmd/dlv/cmds/commands.go:390]: why the capitalization?

File: service/debugger/debugger.go [hunk 10]
```diff
@@ -420,6 +439,20 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
...
+		}
+		err := gobuild.GoBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
```
Comment (aarzilli) [inline service/debugger/debugger.go:450]: This should call GoTestBuild if `Kind == ExecutingGeneratedTest`.

File: service/test/integration2_test.go [hunk 11]
```diff
@@ -74,6 +75,9 @@ func startServer(name string, buildFlags protest.BuildFlags, t *testing.T) (clie
 		Debugger: debugger.Config{
 			Backend:        testBackend,
 			CheckGoVersion: true,
+			Packages:       []string{fixture.Source},
+			BuildFlags:     "", // build flags can be an empty string here because the only test that uses it, does not set special flags.
+			Kind:           debugger.ExecutingGeneratedTest,
```
Comment (aarzilli) [inline service/test/integration2_test.go:80]: This should say ExecutingGeneratedFile, it doesn't really matter but for consistency that's what they are.

File: service/test/integration2_test.go [hunk 12]
```diff
@@ -110,6 +114,7 @@ func TestRunWithInvalidPath(t *testing.T) {
 		APIVersion:  2,
 		Debugger: debugger.Config{
 			Backend: testBackend,
+			Kind:    debugger.ExecutingGeneratedTest,
```
Comment (aarzilli) [inline service/test/integration2_test.go:117]: Also ExecutingGeneratedFile.

File: service/debugger/debugger.go (hunk 6 above)
Comment (alexsaezm) [reply service/debugger/debugger.go]: Thanks for the explanation! I updated the loop to avoid the case

File: service/debugger/debugger.go (hunk 8 above)
Comment (alexsaezm) [reply service/debugger/debugger.go]: Fixed

File: cmd/dlv/cmds/commands.go (hunk 9 above)
Comment (alexsaezm) [reply cmd/dlv/cmds/commands.go]: Because I totally mixed that up :disappointed: . Changed

File: service/test/integration2_test.go (hunk 11 above)
Comment (alexsaezm) [reply service/test/integration2_test.go]: Makes sense. I changed it.

File: service/rpccommon/server.go [hunk 13]
```diff
@@ -116,6 +116,8 @@ func (s *ServerImpl) Run() error {
 		Foreground:           s.config.Foreground,
 		DebugInfoDirectories: s.config.DebugInfoDirectories,
 		CheckGoVersion:       s.config.CheckGoVersion,
+		Packages:             s.config.Packages,
```
[the remark this answers is not in the document]
Comment (alexsaezm) [reply service/rpccommon/server.go]: It should work now :)

File: service/debugger/debugger.go [hunk 14]
```diff
@@ -420,6 +439,21 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 	}
 	var p *proc.Target
 	var err error
+
+	if rebuild {
+		// We cannot build processes that we don't know how to build. If
+		// d.config.Kind is different from ExecutingGeneratedFile that means we
+		// didn't build the binary and we are just running it
+		if d.config.Kind == ExecutingExistingFile {
+			err = gobuild.GoTestBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
```
Comment (aarzilli) [inline service/debugger/debugger.go:448]: This is still wrong:

ExecutingExistingFile, ExecutingOther → it can't be rebuilt at all (either an separately built executable file or a core dump/recording)
ExecutingGeneratedFile → it can be rebuilt by GoBuild
ExecutingGeneratedTest → it can be rebuilt by GoTestBuild

it doesn't make sense to call GoTestBuild on ExecutingExistingFile and it doesn't make sense to call GoBuild on anything other than ExecutingGeneratedFile.

Comment (alexsaezm) [reply service/debugger/debugger.go:451]: Updated. Thanks for the explanation :)

**Comment by aarzilli** [comment]
I'm sorry I forgot to tell you this yesterday but there is a conflict between your PR and something that has already been merged into master that needs to be solved. You need to do a git rebase.

**Comment by alexsaezm** [comment]
I rebased the fork to keep it up to date with master.

File: pkg/terminal/command.go [hunk 15]
```diff
@@ -143,6 +143,7 @@ For live targets the command takes the following forms:
 If newargv is omitted the process is restarted (or re-recorded) with the same argument vector.
 If -noargs is specified instead, the argument vector is cleared.
 `},
+		{aliases: []string{"rebuild"}, group: runCmds, cmdFn: c.rebuild, allowedPrefixes: revPrefix, helpMsg: "Rebuild the code."},
```
Comment (derekparker) [inline pkg/terminal/command.go:146]: We need more documentation here. Add in more information about how it will recompile code, etc, etc...

File: pkg/terminal/command.go [hunk 16]
```diff
@@ -1073,6 +1074,18 @@ func printcontextNoState(t *Term) {
...
 }
 
+func (c *Commands) rebuild(t *Term, ctx callContext, args string) error {
+	if ctx.Prefix == revPrefix {
+		return c.rewind(t, ctx, args)
+	}
+	defer t.onStop()
+	discarded, err := t.client.Restart(true)
+	if len(discarded) > 0 {
+		fmt.Printf("Some of the breakpoints could not be recovered")
```
Comment (derekparker) [inline pkg/terminal/command.go:1084]: nit: no need for capitalization. Also, maybe rephrase to `not all breakpoints could be restored`.

File: service/config.go [hunk 17]
```diff
@@ -36,4 +36,17 @@ type Config struct {
 
 	// DisconnectChan will be closed by the server when the client disconnects
 	DisconnectChan chan<- struct{}
+
+	// TTY is passed along to the target process on creation. Used to specify a
+	// TTY for that process.
+	TTY string
```
Comment (derekparker) [inline service/config.go:42]: This flag can be removed.

File: service/debugger/debugger.go [hunk 18]
```diff
@@ -421,6 +440,26 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
...
+				return nil, fmt.Errorf("could not rebuild process: %s", err)
+			}
+		case ExecutingGeneratedTest:
+			err = gobuild.GoTestBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
+			if err != nil {
+				return nil, fmt.Errorf("could not rebuild process: %s", err)
+			}
+		default:
+			// We cannot build a process that we didn't start, because we don't know
+			// how it was build.
```
Comment (derekparker) [inline service/debugger/debugger.go:458]: s/build/built/

File: service/debugger/debugger.go [hunk 19]
```diff
@@ -450,6 +489,10 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 			}
 			createLogicalBreakpoint(p, addrs, oldBp)
 		} else {
+			// Avoid setting a breakpoint based on address when rebulding
```
Comment (derekparker) [inline service/debugger/debugger.go:492]: s/rebulding/rebuilding/

**Comment by aarzilli** [comment]
> @aarzilli did you have a chance to take another look as well?

I have now. I don't have anything to add. For the rebuild command documentation I think something like this would work:

```
Rebuild the target executable and restarts it. Does not work if the executable was not built by delve.
```

File: cmd/dlv/cmds/commands.go [hunk 20]
```diff
@@ -465,7 +465,7 @@ func traceCmd(cmd *cobra.Command, args []string) {
 		var regexp string
 		var processArgs []string
 
-		dlvArgs, targetArgs := splitArgs(cmd, args)
+		dlvArgs, TargetArgs := splitArgs(cmd, args)
```
Comment (derekparker) [inline cmd/dlv/cmds/commands.go:468]: This does not need to be exported.

File: service/config.go [hunk 21]
```diff
@@ -36,4 +36,13 @@ type Config struct {
 
 	// DisconnectChan will be closed by the server when the client disconnects
 	DisconnectChan chan<- struct{}
+
+	// Packages contains the packages that we are debugging.
```
Comment (derekparker) [inline service/config.go:40]: These three fields can be removed because we've added them to the `debugger.Config` as well, which we include within this config already anyway.

File: service/debugger/debugger.go [hunk 22]
```diff
@@ -98,6 +108,15 @@ type Config struct {
...
 	TTY string
+
+	// Packages contains the packages that we are debugging.
+	Packages []string
+
+	// BuildFlags contains the flags passed to the compiler.
+	BuildFlags string
+
+	// ExecuteKind contains the kind of the executed program.
+	Kind ExecuteKind
```
Comment (derekparker) [inline service/debugger/debugger.go:119]: nit: `Kind` is a little vague, maybe we stutter a bit and name this field `ExecuteKind` as well.

Comment (alexsaezm) [reply service/debugger/debugger.go]: Done. `ExecuteKind` sounds perfect. I updated a few more places where it appears.
